README.md
# HashIDTools
[![Build Status](https://travis-ci.org/joeblackwaslike/hashidtools.svg?branch=master)](https://travis-ci.org/joeblackwaslike/hashidtools) [![Github Repo](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/joeblackwaslike/hashidtools) [![Pypi Version](https://img.shields.io/pypi/v/hashidtools.svg)](https://pypi.python.org/pypi/hashidtools) [![Pypi License](https://img.shields.io/pypi/l/hashidtools.svg)](https://pypi.python.org/pypi/hashidtools) [![Pypi Wheel](https://img.shields.io/pypi/wheel/hashidtools.svg)](https://pypi.python.org/pypi/hashidtools) [![Pypi Versions](https://img.shields.io/pypi/pyversions/hashidtools.svg)](https://pypi.python.org/pypi/hashidtools)


## Maintainer
Joe Black | <me@joeblack.nyc> | [github](https://github.com/joeblackwaslike)


## Introduction
HashID tools/components for async ID generation. Example: `k62K3zOn4Y5Kkxmg7pWOAqPyd8NVjrmX` Designed for async data model ID generation for persisting objects to a graph DB such as ZODB.  ZCA means it's simple to customize and override without editing any code.

#### Comes with
* Customizable Generator/encoder/decoder utility
* Type class (experimental)
* Persistent, 2x BTree-powered IntID/Ref Manager
* Custom fields for `zope.schema` and `attrs`, with default factory functions, validation, etc.

#### Also
* Random seed integer is just under 64bits.
* Derive seed integer at any time by casting type as an int.


## Installation
```shell
pip3 install hashidtools
```

## Usage

### HashIDGenerator
```python
>>> from hashidtools import HashIDGenerator
>>> gen = HashIDGenerator(salt='my random salt', min_length=32)

>>> gen.seed()
...
>>> gen
HashIDGenerator(salt='my random salt', min_length=32)
>>> gen.new()
'...'
>>> gen.decode(gen.new())
...
```


### HashID Type
```python
>>> from hashidtools import HashID
... HashID()
'...'
>>> HashID('8nKqkABjlYB5A7430M917zAJao1Me4mN')
'8nKqkABjlYB5A7430M917zAJao1Me4mN'
```


### Hashid IntID Indexing & Event System
```python
>>> intid = HashIDManager()
... intid
HashIDManager(attribute='id')
... intid.generateId()
'...'

>>> from zc.intid.interfaces import AddedEvent, RemovedEvent
... import zope.event.classhandler
...
>>> @zope.event.classhandler.handler(AddedEvent)
... def handler(event):
...     print(event.id, event.object, event.idmanager)
...
... @zope.event.classhandler.handler(RemovedEvent)
... def handler(event):
...     print(event.id, event.object, event.idmanager)

>>> import attr
... Test = attr.make_class(
...   'Test', {'id': fields.hashid(init=False),
...            'name': attr.ib(default='default-name')})
...

>>> t = Test()
... intid.register(t)
'...'

... id = intid.getId(t)
... intid.getObject(id)
Test(id='...', name='default-name')

>>> intid.unregister(t)
```


### Retrieving the utilities through the ZCA Registry
```python
>>> from zope.component import queryUtility
... from hashidtools.interfaces import IHashIDGenerator, IIntIds
...
... queryUtility(IHashIDGenerator)
HashIDGenerator(salt='$2a$12$AAAAAAAAAAAAAACgpDEPGQ', min_length=32)

>>> queryUtility(IIntIds)
HashIDs(attribute='id')
```


### Extending this Package
This package uses Zope Component Architecture for the ultimate in pluggable extendibility.


#### Quick example of customizing the HashID Generator:
```python
# Note: you can also create your own Generator class that implements the
# IHashIDGenerator interface and register it.
from zope.component import provideUtility

from hashidtools import HashIDGenerator
from hashidtools.interfaces import IHashIDGenerator

generator = HashIDGenerator(**custom_options)
provideUtility(generator, IHashIDGenerator)
```

Note: the following would preferrably be done using your project's ZCML directives.


## Changes
* [CHANGELOG](CHANGELOG.md)


CHANGELOG.md
# HashIDTools Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2018-05-08
### Added
- Initial commit.


tests/test_types.py
import unittest

import attr
from zope import component
import zope.schema
from zc.intid.interfaces import IIntIds, AddedEvent, RemovedEvent
import zope.event.classhandler

import hashidtools
from hashidtools import fields
from hashidtools.interfaces import IHashIDGenerator, IHashID
from hashidtools.types import HashIDGenerator, HashID, HashIDManager


class TestHashIDGenerator(unittest.TestCase):
    def makeOne(self, salt='sdfs', min_length=32):
        return HashIDGenerator(salt=salt, min_length=min_length)

    def test_interface(self):
        gen = self.makeOne()
        self.assertTrue(IHashIDGenerator.providedBy(gen))

    def test_schema(self):
        gen = self.makeOne()
        errors = zope.schema.getSchemaValidationErrors(IHashIDGenerator, gen)
        self.assertEqual(errors, [])

    def test_hashid_generator_basic(self):
        salt = 'sdfs'
        min_length = 34
        gen = self.makeOne(salt, min_length)
        self.assertEqual(gen.salt, salt)
        self.assertEqual(gen.min_length, min_length)

    def test_hashid_generator_seed(self):
        gen = self.makeOne()
        seed = gen.seed()
        self.assertIsInstance(seed, int)
        bit_length = int(seed).bit_length()
        self.assertGreater(bit_length, 50)
        self.assertLess(bit_length, 64)

    def test_hashid_generator_encode_decode(self):
        gen = self.makeOne()
        seed = 1762352222709391612
        hashid = gen.encode(seed)

        self.assertIsInstance(hashid, str)
        self.assertEqual(hashid, 'bNo4jLpM8mK2PW6l4260YGXnyQlBE1k3')
        self.assertEqual(seed, gen.decode(hashid))

    def test_hashid_generator_new(self):
        gen = self.makeOne()
        hashid = gen.new()
        self.assertEqual(len(hashid), 32)
        self.assertRegex(hashid, r'^\w{32}$')
        self.assertIsInstance(hashid, str)

    def test_immutable_attributes(self):
        from attr.exceptions import FrozenInstanceError

        gen = self.makeOne()
        with self.assertRaises(FrozenInstanceError):
            gen.salt = 'new'

        with self.assertRaises(FrozenInstanceError):
            gen.min_length = 43

        with self.assertRaises(FrozenInstanceError):
            gen.seed_bits = 32



class TestHashID(unittest.TestCase):
    def makeOne(self, id=None):
        if id:
            return HashID(id)
        else:
            return HashID()

    def test_interface(self):
        hashid = self.makeOne()
        self.assertTrue(IHashID.providedBy(hashid))

    def test_schema(self):
        hashid = self.makeOne()
        errors = zope.schema.getSchemaValidationErrors(IHashID, hashid)
        self.assertEqual(errors, [])

    def test_hashid_provided(self):
        id = '5MopkXVL7Ej9dWkaR9kBqNGYway3RAbd'
        hashid = self.makeOne(id)
        self.assertIsInstance(hashid, HashID)
        self.assertEqual(len(hashid), 32)
        self.assertEqual(hashid.id, id)

    def test_hashid_generated(self):
        hashid = self.makeOne()
        self.assertIsInstance(hashid, HashID)
        self.assertEqual(len(hashid), 32)

    def test_hashid_reprs(self):
        hashid = self.makeOne()
        self.assertEqual(repr(hashid), '{!r}'.format(hashid.id))
        self.assertTrue(str(hashid), '{!s}'.format(hashid.id))

    def test_hashid_int_cast(self):
        gen = HashIDGenerator()
        seed = gen.seed()
        hashid = HashID(gen.encode(seed))
        self.assertEqual(int(hashid), seed)

    def test_hashid_dictionary_key(self):
        id1 = self.makeOne()
        id2 = self.makeOne()
        d = {
            id1: 'value',
            id2: 'value2'
        }
        self.assertEqual(d[id1], d[str(id1)])
        self.assertEqual(d[id2], d[str(id2)])

    def test_immutable_attributes(self):
        from attr.exceptions import FrozenInstanceError

        hashid = self.makeOne()
        with self.assertRaises(FrozenInstanceError):
            hashid.id = 'new'


@attr.s
class Fixture:
    id: str = fields.hashid(init=False)
    name: str = attr.ib(default='default-name')


class TestHashIDManager(unittest.TestCase):
    def makeOne(self):
        return HashIDManager()

    def test_interface(self):
        intid = self.makeOne()
        self.assertTrue(IIntIds.providedBy(intid))

    def test_hashid_provided(self):
        inst = Fixture('test')

        intid = self.makeOne()
        hashid = intid.generateId(inst)

        self.assertRegex(hashid, r'^\w{32}$')

        _hid = intid.getId(inst)
        self.assertEqual(_hid, str(inst.id))

        added = []
        @zope.event.classhandler.handler(AddedEvent)
        def added_handler(event):
            added.append(event)
            print(f'id: {event.id} obj:{event.object} intid:{event.idmanager}')

        removed = []
        @zope.event.classhandler.handler(RemovedEvent)
        def removed_handler(event):
            removed.append(event)
            print(f'id: {event.id} obj:{event.object} intid:{event.idmanager}')

        _hid = intid.register(inst)
        self.assertEqual(_hid, str(inst.id))
        self.assertEqual(len(added), 1)

        intid.unregister(inst)
        self.assertEqual(len(removed), 1)


tests/test_fields.py
import unittest

import attr
from zope.interface import Interface, implementer
import zope.schema

import hashidtools
from hashidtools.interfaces import IHashIDGenerator, IHashID
from hashidtools import HashIDGenerator, HashID
from hashidtools import fields


class IFoo(Interface):
    id = fields.HashID()

@implementer(IFoo)
@attr.s(auto_attribs=True)
class Foo:
    id: HashID = None


@implementer(IFoo)
@attr.s(auto_attribs=True)
class Foo2:
    id: HashID = fields.hashid()


class TestHashID(unittest.TestCase):
    def makeFoo(self, *args, **kwargs):
        return Foo(*args, **kwargs)

    def makeFoo2(self, *args, **kwargs):
        return Foo2(*args, **kwargs)

    # def test_schema_field(self):
    #     foo = self.makeFoo()
    #
    #     for name, field in zope.schema.getFields(IFoo).items():
    #         bound = field.bind(foo)
    #         bound.set(foo, name)
    #
    #         bound.validate(bound.get(foo))
    #
    #         errors = zope.schema.getValidationErrors(IFoo, foo)
    #         self.assertEqual(errors, [])

    # def test_attrs_field(self):
    #     foo = self.makeFoo2()
    #
    #     for name, field in zope.schema.getFields(IFoo).items():
    #         bound = field.bind(foo)
    #         bound.set(foo, name)
    #
    #         bound.validate(bound.get(foo))
    #
    #         errors = zope.schema.getValidationErrors(IFoo, foo)
    #         self.assertEqual(errors, [])


hashidtools/fields.py
"""
hashidtools.fields
~~~~~~~~~~~~~~~~

Fields helpers for zope.schema and attrs.

:copyright: (c) 2018 by Joseph Black.
:license: MIT, see LICENSE for more details.
"""

import re

from zope.interface import implementer
from zope.schema.interfaces import IFromUnicode
from zope.schema._field import NativeStringLine
from zope.component import queryUtility

import attr
from attr.validators import instance_of
from .interfaces import IHashID, IHashIDGenerator
from .types import HashID as HashIDType
from .exceptions import InvalidHashID

HASHID_REGEX = re.compile(r'^\w{32}$')


@implementer(IHashID, IFromUnicode)
class HashID(NativeStringLine):
    """HashID field for zope.schema.

    Values of HashID fields must be a alphanum string 32chars in length.
    """

    def __init__(self, *args, **kwargs):
        kwargs.setdefault('title', 'HashID')
        kwargs.setdefault(
            'description', 'A short string representing the HashID.')
        kwargs.setdefault('required', True)
        kwargs.setdefault('readonly', True)
        kwargs.setdefault('constraint', HASHID_REGEX.match)
        kwargs.setdefault(
            'defaultFactory',
            lambda: queryUtility(IHashIDGenerator).new())
        super(HashID, self).__init__(*args, **kwargs)

    def _validate(self, value):
        super(HashID, self)._validate(value)
        if HASHID_REGEX.match(value):
            return
        raise InvalidHashID(value)

    def fromUnicode(self, value):
        """See IFromUnicode."""
        v = value.strip()
        if isinstance(v, bytes):
            v = v.decode()
        v = HashIDType(v)
        self.validate(v)
        return v


def hashid(*args, **kwargs):
    """HashID field for attrs."""
    return attr.ib(
        validator=[instance_of(str)],
        factory=queryUtility(IHashIDGenerator),
        **kwargs)


hashidtools/__init__.py
"""
hashidtools
~~~~~~~~~

HashID tools/components for async ID generation.

Designed for async data model ID generation for persisting objects to a graph
DB such as ZODB.  ZCA means it's simple to customize without editing any code.

Example: `k62K3zOn4Y5Kkxmg7pWOAqPyd8NVjrmX`

Usage::

    >>> from zc.intid.interfaces import IIntIds
    ... from zope.component import queryUtility
    ... from hashidtools.interfaces import IHashIDGenerator, IHashID, IHashID
    ...
    ... queryUtility(IHashIDGenerator)
    HashIDGenerator(salt='$2a$12$AAAAAAAAAAAAAACgpDEPGQ', min_length=32)

    >>> queryUtility(IHashIDGenerator)()
    '...'

    >>> queryUtility(IIntIds)
    HashIDManager(attribute='id')

    >>> intid = queryUtility(IIntIds)
    ... intid
    HashIDManager(attribute='id')

    >>> from zc.intid.interfaces import AddedEvent, RemovedEvent
    ... import zope.event.classhandler
    ...
    ... @zope.event.classhandler.handler(AddedEvent)
    ... def handler(event):
    ...     print(event.id, event.obj, event.idmanager)
    ...
    ... @zope.event.classhandler.handler(RemovedEvent)
    ... def handler(event):
    ...     print(event.id, event.obj, event.idmanager)

    >>> import attr
    ... Test = attr.make_class(
    ...   'Test', {'id': fields.hashid(init=False),
    ...            'name': attr.ib(default='default-name')})

    >>> t = Test()
    ... intid.register(t)
    '...'

    ... id = intid.getId(t)
    ... intid.getObject(id)
    Test(id='...', name='default-name')

    >>> intid.unregister(t)

:copyright: (c) 2018 by Joseph Black.
:license: MIT, see LICENSE for more details.
"""


__version__ = '1.0.2'
__title__ = "hashidtools"

from zope.configuration import xmlconfig

from . import interfaces, exceptions, types, fields
from .types import HashIDGenerator, HashID, HashIDManager


xmlconfig.file('configure.zcml', __import__('sys').modules[__name__])


hashidtools/interfaces.py
# pylint: disable=inherit-non-class,no-self-argument,no-method-argument
# pylint: disable=unexpected-special-method-signature,arguments-differ

"""
hashidtools.interfaces
~~~~~~~~~~~~~~~~

Zope compatible interfaces and schemata for this package.

:copyright: (c) 2018 by Joseph Black.
:license: MIT, see LICENSE for more details.
"""

import re

from zope.interface import Interface
import zope.schema
from zope.component import queryUtility


class IHashIDAware(Interface):
    """Marker for interfaces that support HashIDs."""


class IHashIDGenerator(Interface):
    """Generator of HashID encoded ID's."""

    salt = zope.schema.TextLine(
        title='salt', description="a short string to use as crypto salt",
        readonly=True, required=True, default='$2a$12$AAAAAAAAAAAAAACgpDEPGQ'
    )
    min_length = zope.schema.Int(
        title='minimum id length',
        description="The minimum length for the generated hashid",
        readonly=True, required=True, default=32, min=32, max=128
    )
    alphabet = zope.schema.TextLine(
        title='Hashid alphabet',
        description="A custom alphabet to use when generating hashids",
        readonly=True,
        default=('abcdefghijklmnopqrstuvwxyz'
                 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890')
    )

    def __call__():
        """Shortcut to new()"""

    def seed():
        """Return a randomly generated ~64bit int seed."""

    def new():
        """Return a new hashid value."""

    def decode(hashid):
        """Decode a hashid value to it's base integer."""


class IHashID(IHashIDAware):
    """HashID type of 64bit integer, used for ZODB object ID generation."""

    id = zope.schema.TextLine(
        title='HashID', description="A HashIDs based ID.",
        readonly=True, required=True, constraint=re.compile(r'^\w{32}$').match,
        defaultFactory=queryUtility(IHashIDGenerator))

    def __int__():
        """Return integer representation of the HashID value."""


hashidtools/types.py
"""
hashidtools.hashid
~~~~~~~~~~~~~~~~

Miscellaneous HashID Tools

:copyright: (c) 2018 by Joseph Black.
:license: MIT, see LICENSE for more details.
"""

import random
from typing import ClassVar, Union
from weakref import WeakKeyDictionary

import hashids
from zope.interface import implementer
from zope.component import queryUtility
from zope.event import notify
from zope.intid.interfaces import IntIdMissingError
from zope.security.proxy import removeSecurityProxy as unwrap
from zc.intid.interfaces import (
    IIntIds, IntIdInUseError, AddedEvent, RemovedEvent)
from zc.intid.utility import IntIds
import BTrees
import attr
from attr.validators import instance_of

from .interfaces import IHashIDGenerator, IHashID
from .exceptions import InvalidHashID, IDRegisterError


@implementer(IHashIDGenerator)
@attr.s(frozen=True)
class HashIDGenerator:
    """Generator of HashID encoded ID's.

    :param salt str: A short string to use as the unique salt.
    :param min_length int: (32) The minimum length of of the generated HashID.
    :return: a HashIDGenerator object.
    :rtype: :inst:`HashIDGenerator`

    Usage::

        >>> from hashidtools import HashIDGenerator
        >>> HashIDGenerator(salt='my random salt', min_length=32)

        >>> from hashidtools.interfaces import IHashIDGenerator
        ... from zope.component import queryUtility
        ... queryUtility(IHashIDGenerator)
        HashIDGenerator(salt='$2a$12$AAAAAAAAAAAAAACgpDEPGQ', min_length=32)

        >>> queryUtility(IHashIDGenerator)()
        '...'
        >>> queryUtility(IHashIDGenerator).decode('k62K3zOn4Y5Kkxmg7pWOAqPyd8NVjrmX')
        1032596908023458124
    """

    salt: str = attr.ib(
        default='$2a$12$AAAAAAAAAAAAAACgpDEPGQ',
        converter=str,
        validator=instance_of(str))
    min_length: int = attr.ib(
        default=32,
        converter=int,
        validator=instance_of(int))
    alphabet: str = attr.ib(
        default=('abcdefghijklmnopqrstuvwxyz'
                 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
                 '1234567890'),
        validator=instance_of(str),
        repr=False
    )

    def __attrs_post_init__(self):
        super(HashIDGenerator, self).__setattr__(
            '_gen', hashids.Hashids(self.salt, self.min_length, self.alphabet))

    def __call__(self):
        return self.new()

    def seed(self):
        """Return a randomly generated ~64bit int seed."""
        return random.getrandbits(64-1)

    def encode(self, value):
        """HashID encode an integer value."""
        return self._gen.encode(value)

    def new(self, seed=None):
        """Return a new hashid value."""
        seed = seed or self.seed()
        return self.encode(seed)

    def decode(self, hashid):
        """Decode a hashid value to it's base integer."""
        return self._gen.decode(hashid)[0]


@implementer(IHashID)
@attr.s(frozen=True, hash=False, repr=False, cmp=False)
class HashID:
    """HashID type of 64bit integer, used for ZODB object ID generation.

    :param id str: A short string representing the HashID.
    :return: A HashID object.
    :rtype: :inst:`HashID`

    Usage::

        >>> from hashidtools import HashID
        >>> HashID()
        '...'
        >>> HashID('8nKqkABjlYB5A7430M917zAJao1Me4mN')
        '8nKqkABjlYB5A7430M917zAJao1Me4mN'

    """

    id = attr.ib(validator=instance_of(str))

    @id.default
    def _default_id(self):
        return queryUtility(IHashIDGenerator)()

    def __hash__(self):
        return hash(str(self.id))

    def __len__(self):
        return len(self.id)

    def __eq__(self, other):
        if isinstance(other, str):
            return str(self.id) == str(other)
        return str(self.id) == str(other.id)

    def __gt__(self, other):
        if isinstance(other, str):
            return str(self.id) > str(other)
        return str(self.id) > str(other.id)

    def __lt__(self, other):
        if isinstance(other, str):
            return str(self.id) < str(other)
        return str(self.id) < str(other.id)

    def __repr__(self):
        return repr(self.id)

    def __str__(self):
        return str(self.id)

    def __int__(self):
        return queryUtility(IHashIDGenerator).decode(self.id)


# https://media.readthedocs.org/pdf/zopecatalog/latest/zopecatalog.pdf
@implementer(IIntIds)
@attr.s
class HashIDManager(IntIds):
    """HashID IntId Manager.

    :param BTrees.family type: A short string representing the HashID.
    :return: A HashIDManager object.
    :rtype: :inst:`HashIDManager`

    Usage::

        >>> intid = HashIDManager()
        ... intid
        HashIDManager(attribute='id')
        ... intid.generateId()
        '...'

        >>> from zc.intid.interfaces import AddedEvent, RemovedEvent
        ... import zope.event.classhandler
        ...
        >>> @zope.event.classhandler.handler(AddedEvent)
        ... def handler(event):
        ...     print(event.id, event.object, event.idmanager)
        ...
        ... @zope.event.classhandler.handler(RemovedEvent)
        ... def handler(event):
        ...     print(event.id, event.object, event.idmanager)

        >>> import attr
        ... Test = attr.make_class(
        ...   'Test', {'id': fields.hashid(init=False),
        ...            'name': attr.ib(default='default-name')})
        ...

        >>> t = Test()
        ... intid.register(t)
        '...'

        ... id = intid.getId(t)
        ... intid.getObject(id)
        Test(id='...', name='default-name')

        >>> intid.unregister(t)

    """

    family: ClassVar[type] = BTrees.family64
    attribute: str = attr.ib(
        default='id',
        validator=instance_of(str))

    def __attrs_post_init__(self):
        self.ids = self.family.OO.BTree()
        self.refs = self.family.OO.BTree()

    def generateId(self, obj=None):
        """Generate ID."""
        return queryUtility(IHashIDGenerator)()

    def getId(self, obj):
        """Return the ID for passed object."""
        unwrapped = unwrap(obj)
        uid = getattr(unwrapped, self.attribute, None)
        if not isinstance(uid, str):
            uid = str(uid)
        if uid is None:
            raise IntIdMissingError(obj)
        return uid

    def register(self, obj):
        """Register objects to ID."""
        obj = unwrap(obj)
        uid = self.queryId(obj)
        if uid is None:
            uid = self.generateId(obj)
            if uid in self.refs:
                raise IntIdInUseError("id generator returned used id")
        if uid != getattr(obj, self.attribute):
            raise IDRegisterError(f'uid: {uid} != obj.id: {obj.id}')
        self.refs[uid] = obj
        notify(AddedEvent(obj, self, uid))
        return uid
